eessi/testsuite/tests/apps/osu.py

```diff
@@ -0,0 +1,314 @@
+        # required since the non CUDA module should be able to run in the GPU
+        # partition as well. This is specific for this test and not covered by
+        # the function above.
+        if not is_cuda_module and self.device_type == DEVICE_TYPES[GPU]:
+            self.valid_systems = [f'+{FEATURES[GPU]} %{GPU_VENDOR}={GPU_VENDORS[NVIDIA]}']
+            self.device_buffers = 'cpu'
+        elif is_cuda_module and self.device_type == DEVICE_TYPES[GPU]:
+            # Currently the device buffer is hard coded to be cuda. More
+            # options need to be introduced based on vendor and device type.
+            self.device_buffers = 'cuda'
+
+        # If the device_type is CPU then device buffer should always be CPU.
+        if self.device_type == DEVICE_TYPES[CPU]:
+            self.device_buffers = 'cpu'
+
+        # This part of the code removes the collective communication calls out
+        # of the run list since this test is only meant for pt2pt.
+        if not self.benchmark_info[0].startswith('mpi.pt2pt'):
+            self.valid_systems = []
+        hooks.set_modules(self)
+
+    @run_after('init')
+    def set_tag_ci(self):
+        """ Setting tests under CI tag. """
+        if (self.benchmark_info[0] in ['mpi.pt2pt.osu_latency',
+                                       'mpi.pt2pt.osu_bw']):
+            self.tags.add('CI')
+            log(f'tags set to {self.tags}')
+
+        if (self.benchmark_info[0] == 'mpi.pt2pt.osu_bw'):
+            self.tags.add('osu_bw')
+
+        if (self.benchmark_info[0] == 'mpi.pt2pt.osu_latency'):
+            self.tags.add('osu_latency')
+
+    @run_after('init')
+    def set_mem(self):
+        """ Setting an extra job option of memory. This test has only 4
+        possibilities: 1_node, 2_nodes, 2_cores and 1_cpn_2_nodes. Only the
+        last 2 require the memory to be set. """
+        is_cuda_module = utils.is_cuda_required_module(self.module_name)
+        if(SCALES.get(self.scale).get('node_part', 0) == 0):
+            self.extra_resources = {'memory': {'size': '32GB'}}
+
+    @run_after('init')
+    def set_num_tasks(self):
+        """ Setting scales as tags. """
+        hooks.set_tag_scale(self)
+
+    @run_after('init')
+    def set_num_tasks_per_node(self):
+        if(SCALES.get(self.scale).get('num_nodes') == 1):
+            self.num_tasks_per_node = 2
+        else:
+            self.num_tasks_per_node = 1
+
+    @run_after('setup')
+    def set_num_cpus_per_task(self):
+        """ Since num_tasks_per_node is already set. This function sets
+        num_cpus_per_task for 1 node and 2 node options where the request is
+        for full nodes."""
+        if(SCALES.get(self.scale).get('num_nodes') >= 1 and
+           SCALES.get(self.scale).get('node_part', 0) == 1):
+            hooks.assign_one_task_per_compute_unit(self,
+                                                   COMPUTE_UNIT.get(CPU, 'cpu'))
```

Comment:
you should not use `assign_one_task_per_compute_unit`, because for the pt2pt test we want
- for 2 nodes: 1 task per **node**, which is not implemented yet (we could implement this though)
- for 1 node: 2 tasks per node

that's the reason you should set the resources yourself, making use of `assign_default_cpus_per_node`

Reply:
@satishskamath i've added an extra hook in https://github.com/EESSI/test-suite/pull/97

with the new hook you can use:
`hooks.assign_tasks_per_compute_unit(self, COMPUTE_UNIT[NODE])` for 1 task per node, and
`hooks.assign_tasks_per_compute_unit(self, COMPUTE_UNIT[NODE], num_per=2)` for 2 tasks per node.

Reply:
Hi @smoors,

Even with the old code, for 2 full nodes, I get 1 task per node.
This one is on a `gpu` node doing `cpu` tests.
```
#!/bin/bash
#SBATCH --job-name="rfm_job"
#SBATCH --ntasks=2
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=72
#SBATCH --output=rfm_job.out
#SBATCH --error=rfm_job.err
#SBATCH --time=0:30:0
#SBATCH -p gpu
#SBATCH --export=None
#SBATCH --gpus-per-node=4
module load 2022
module load OSU-Micro-Benchmarks/5.9-gompi-2022a
mpirun -np 2 osu_latency -m 8 -x 10 -i 1000 -c D D
```
This one on a `cpu` partition with scale `2_nodes`:
```
#!/bin/bash
#SBATCH --job-name="rfm_job"
#SBATCH --ntasks=2
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=128
#SBATCH --output=rfm_job.out
#SBATCH --error=rfm_job.err
#SBATCH --time=0:30:0
#SBATCH -p thin
#SBATCH --export=None
module load 2022
module load OSU-Micro-Benchmarks/5.9-gompi-2022a
mpirun -np 2 osu_latency -m 8 -x 10 -i 1000 -c D D
```

Reply:
If you see line 108:
```
    @run_after('init')
    def set_num_tasks_per_node(self):
        if(SCALES.get(self.scale).get('num_nodes') == 1):
            self.num_tasks_per_node = 2
        else:
            self.num_tasks_per_node = 1

```
I set the `num_tasks_per_node` there.

Reply:
that's not a clean way to do this.
`hooks.assign_one_task_per_compute_unit(self, COMPUTE_UNIT[CPU])` is only meant to set 1 task per cpu core,
so manually setting `self.num_tasks_per_node = 1` is confusing and it does not take into account if a user wants to set custom values.

Reply:
Well in this case, I thought the user is not supposed to set the `num_tasks_per_node` at all. Based on the scale selected the test should automatically set it and the user should not be able to set custom values in this case. That is the reason it is hard coded. But if you think that it is cleaner to do it in that way, then I can change it as well.

Reply:
i understand, but i don't see why we should limit users from trying crazy things, as long as we are setting optimal resources by default. that's what we also do for the other tests, so it would be confusing if we did it differently here.

Reply:
Ok. :) Doesn't hurt to keep things consistent. I will make the change.

Reply:
Change incorporated in setting `num_tasks_per_node`.